File: android/src/test/kotlin/com/hydrapayments/sdk/ExceptionTests.kt
package com.hydrapayments.sdk

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ExceptionTests {
    @Test
    fun hydraErrorDefaults() {
        val error = HydraError("Something went wrong")
        assertEquals("Something went wrong", error.message)
        assertEquals(500, error.statusCode)
        assertEquals("API_ERROR", error.errorCode)
        assertNull(error.details)
    }

    @Test
    fun hydraErrorWithCustomErrorCodeAndDetails() {
        val details = mapOf("field" to "amount")
        val error = HydraError("Custom error", 422, "CUSTOM_ERROR", details)
        assertEquals(422, error.statusCode)
        assertEquals("CUSTOM_ERROR", error.errorCode)
        assertEquals("amount", error.details?.get("field"))
    }

    @Test
    fun authenticationError() {
        val error = AuthenticationError("Invalid API key")
        assertEquals(401, error.statusCode)
        assertEquals("AUTHENTICATION_ERROR", error.errorCode)
    }

    @Test
    fun validationError() {
        val error = ValidationError("Validation failed")
        assertEquals(400, error.statusCode)
        assertEquals("VALIDATION_ERROR", error.errorCode)
    }

    @Test
    fun notFoundError() {
        val error = NotFoundError("Not found")
        assertEquals(404, error.statusCode)
        assertEquals("NOT_FOUND", error.errorCode)
    }

    @Test
    fun allSubclassesAreHydraError() {
        val errors: List<HydraError> = listOf(
            AuthenticationError("a"),
            ValidationError("v"),
            NotFoundError("n"),
            HydraError("h")
        )
        assertEquals(4, errors.size)
        errors.forEach { assertTrue(it is HydraError) }
    }

    @Test
    fun detailsPassedToSubclass() {
        val details = mapOf("reason" to "test")
        val error = ValidationError("bad", details)
        assertEquals("test", error.details?.get("reason"))
    }
}

File: ios/Tests/HydraPaymentsTests/ExceptionTests.swift
import XCTest
@testable import HydraPayments

final class ExceptionTests: XCTestCase {
    func testHydraErrorDefaults() {
        let error = HydraError(message: "Something went wrong")
        XCTAssertEqual(error.message, "Something went wrong")
        XCTAssertEqual(error.statusCode, 500)
        XCTAssertEqual(error.errorCode, "API_ERROR")
        XCTAssertNil(error.details)
    }

    func testHydraErrorWithCustomErrorCodeAndDetails() {
        let details = ["field": "amount"]
        let error = HydraError(message: "Custom error", statusCode: 422, errorCode: "CUSTOM_ERROR", details: details)
        XCTAssertEqual(error.statusCode, 422)
        XCTAssertEqual(error.errorCode, "CUSTOM_ERROR")
        XCTAssertEqual(error.details?["field"], "amount")
    }

    func testAuthenticationError() {
        let error = AuthenticationError(message: "Invalid API key")
        XCTAssertEqual(error.statusCode, 401)
        XCTAssertEqual(error.errorCode, "AUTHENTICATION_ERROR")
    }

    func testValidationError() {
        let error = ValidationError(message: "Validation failed")
        XCTAssertEqual(error.statusCode, 400)
        XCTAssertEqual(error.errorCode, "VALIDATION_ERROR")
    }

    func testNotFoundError() {
        let error = NotFoundError(message: "Not found")
        XCTAssertEqual(error.statusCode, 404)
        XCTAssertEqual(error.errorCode, "NOT_FOUND")
    }

    func testAllSubclassesAreHydraError() {
        let errors: [HydraError] = [
            AuthenticationError(message: "a"),
            ValidationError(message: "v"),
            NotFoundError(message: "n"),
            HydraError(message: "h")
        ]
        XCTAssertEqual(errors.count, 4)
        for error in errors {
            XCTAssertEqual(error.statusCode > 0, true)
        }
    }

    func testDetailsPassedToSubclass() {
        let details = ["reason": "test"]
        let error = ValidationError(message: "bad", details: details)
        XCTAssertEqual(error.details?["reason"], "test")
    }
}
